### PR TITLE
[TD]multiThread Release Candidate 1 - DO NOT MERGE

### PR DESCRIPTION
**Pull request by WandererFan**
This PR contains the initial version of using separate threads for HLR and face finding in TD.  It is for testing only,  Do not merge!

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [ ]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [ ]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [592597651](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/592597651) was triggered at [b8b2da0](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/b8b2da031ca8938ad1da1bab1e193ff19cf9aed7). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/TechDraw/Gui/QGSPage.cpp [hunk 1]
```diff
@@ -505,26 +698,95 @@ QGIView * QGSPage::addWeldSymbol(TechDraw::DrawWeldSymbol* weld)
...
+{
+//    Base::Console().Message("QGSP::setLeaderGroups()\n");
+    const std::vector<QGIView *> &allItems = getViews();
+    std::vector<QGIView *>::const_iterator itInspect;
+    int leadItemType = QGraphicsItem::UserType + 232;
+
+    //make sure that qgileader belongs to correct parent.
+    //quite possibly redundant
+    for (itInspect = allItems.begin(); itInspect != allItems.end(); itInspect++) {
+        if (((*itInspect)->type() == leadItemType) && (!(*itInspect)->group())) {
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/QGSPage.cpp:750]:
```suggestion
        if ((*itInspect)->type() == leadItemType && !(*itInspect)->group()) {
```
If possible, also avoid nesting. If possible, all these code parts should be gathered in a common function. Let's avoid shotgun surgery (https://en.wikipedia.org/wiki/Shotgun_surgery)

File: src/Mod/TechDraw/Gui/QGSPage.cpp [hunk 2]
```diff
@@ -505,26 +698,95 @@ QGIView * QGSPage::addWeldSymbol(TechDraw::DrawWeldSymbol* weld)
...
+void QGSPage::setRichAnnoGroups(void)
+{
+//    Base::Console().Message("QGSP::setRichAnnoGroups()\n");
+    const std::vector<QGIView *> &allItems = getViews();
+    std::vector<QGIView *>::const_iterator itInspect;
+    int annoItemType = QGraphicsItem::UserType + 233;
+
+    //make sure that qgirichanno belongs to correct parent.
+    for (itInspect = allItems.begin(); itInspect != allItems.end(); itInspect++) {
+        if (((*itInspect)->type() == annoItemType) && (!(*itInspect)->group())) {
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/QGSPage.cpp:769]:
```suggestion
        if ((*itInspect)->type() == annoItemType && !(*itInspect)->group()) {
```
If possible, also avoid nesting. If possible, all these code parts should be gathered in a common function. Let's avoid shotgun surgery (https://en.wikipedia.org/wiki/Shotgun_surgery)

File: src/Mod/TechDraw/Gui/QGSPage.cpp [hunk 3]
```diff
@@ -585,6 +848,36 @@ QGIView *QGSPage::findParent(QGIView *view) const
...
+       App::DocumentObject* obj = lead->LeaderParent.getValue();
+       if(obj != nullptr) {
+           std::string parentName = obj->getNameInDocument();
+           for(std::vector<QGIView *>::const_iterator it = qviews.begin(); it != qviews.end(); ++it) {
+               if(strcmp((*it)->getViewName(), parentName.c_str()) == 0) {
+                   return *it;
+               }
+           }
+       }
+   }
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/QGSPage.cpp:865]: Wow, here we should really avoid nesting.

File: src/Mod/TechDraw/Gui/QGSPage.cpp [hunk 4]
```diff
@@ -585,6 +848,36 @@ QGIView *QGSPage::findParent(QGIView *view) const
...
+       App::DocumentObject* obj = anno->AnnoParent.getValue();
+       if (obj != nullptr) {
+           std::string parentName = obj->getNameInDocument();
+           for(std::vector<QGIView *>::const_iterator it = qviews.begin(); it != qviews.end(); ++it) {
+               if(strcmp((*it)->getViewName(), parentName.c_str()) == 0) {
+                   return *it;
+               }
+           }
+       }
+   }
```
Comment (benj5378) [inline src/Mod/TechDraw/Gui/QGSPage.cpp:879]: Wow, here we should really avoid nesting.

**Review by benj5378 — COMMENTED** [review]
Most of my comments might relate to moved code that you didn't write for this PR. If this is the case, ignore the the comments and let's solve it in a seperate PR. Only if newly written code, take note of my comments

**Comment by WandererFan** [comment]
Let's get it working before we make it pretty.  With all the mods going on now this code is going to have to be heavily modified just to get it to merge.

**Comment by donovaly** [comment]
Many thanks for working on this!

Can you please rebase the PR (to save compile time) and give here or in the forum some info how to test?

(p.s. I see you re-add some nullptr checks I just removed this week. As we discussed, they are not necessary and should be avoided.)

**Comment by WandererFan** [comment]
The latest version is rebased to today's master.  

What discussion was that?

**Comment by donovaly** [comment]
> The latest version is rebased to today's master.

Thanks. Do you have any hints for me how to test? Or should I just take some of my real-life documents?

> What discussion was that?

I remember several during the last year. The recent one I referred to is this:
https://github.com/FreeCAD/FreeCAD/pull/7202#discussion_r922876606

**Comment by WandererFan** [comment]
To test, you just need a file that takes a long time to load/recompute.  Preferably one that has multiple views, as the only difference with a single view will be that the Gui should be unlocked for most of the waiting time.

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [600449008](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/600449008) was triggered at [b128a91](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/b128a916b3a95f0e9eb64e895e8f069ac1be4e38). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [600567110](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/600567110) was triggered at [24448c6](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/24448c60cc0ba38501b42aa528527e25a6e969f8). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [603291313](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/603291313) was triggered at [6159fba](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/6159fbafd202429ff1cffca4fd5f93c2407accab). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [606119510](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/606119510) was triggered at [c4974b8](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/c4974b8b3a48c5a176a74632b1070f50928e11ab). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
New snap build has been triggered

**Comment by luzpaz** [comment]
A build from commit 6af2d9f15d1eeba84a36753ad2186f9dab5b8928 has been published to channel edge/pr-44 as revision 192.

  ## Installation instructions

  ```bash
    # Note: not all distros require 'sudo'
    $ sudo snap install --channel=edge/pr-44 freecad

    # If you already have the freecad snap installed
    $ sudo snap refresh --channel=edge/pr-44 freecad 

    # Experimental: Parallel install (see https://github.com/FreeCAD/FreeCAD-snap/blob/master/docs/index.md#parallel-installs)
    $ sudo snap set system experimental.parallel-instances=true
    $ sudo snap install --channel=edge/pr-44 freecad_pr44
    $ freecad_pr44 # Run it

    # To switch back to preferred snap (swap out 'edge' for 'stable' if necessary)
    $ snap refresh freecad --channel=edge freecad
  ```

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [607823760](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/607823760) was triggered at [72490d8](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/72490d8a2659abbad6526ea7ea269240e1d49791). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
Snap package has been re-triggered. New build is being generate and will be available soon from https://github.com/FreeCAD/FreeCAD-snap/pull/44

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [609880039](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/609880039) was triggered at [76c446f](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/76c446f5cecfc32228ed54017842214681ad5032). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [610707778](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/610707778) was triggered at [7a40e00](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/7a40e0026b490fd99743c9e3edee78b5f757a28f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by berndhahnebach** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224">pipeline status</a> for feature branch [PR_7224](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_7224). Pipeline [612026275](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/612026275) was triggered at [0df187d](https://github.com/FreeCAD/FreeCAD/pull/7224/commits/0df187d22876bc2e5516ff86d52b1da3a74938ff). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Pull request closed without merging**